### PR TITLE
[WIP] Update polyhedron(::JuMP.Model) to Jump 0.19

### PR DESCRIPTION
**Pull request by OliverEvans96**
Hi all,

Thanks for the great package! I've been working on updating the LP -> polyhedron conversion method to JuMP 0.19 (solves #152 in part). It seems to be working for the simple case I've dealt with, though I haven't yet written proper tests so far.

I'm additionally having a bit of a versioning issue, #154, which is preventing me from running tests properly.

Any feedback and suggestions are appreciated!

Cheers,
Oliver

### CODE REVIEW

**Comment by blegat** [comment]
> I'm additionally having a bit of a versioning issue, #154, which is preventing me from running tests properly.

You need to update the REQUIRE file to `JuMP 0.19`

File: src/jump.jl [hunk 1]
```diff
@@ -8,18 +8,143 @@ Note that if non-linear constraint are present in the model, they are ignored.
 """
 hrep(model::JuMP.Model) = LPHRepresentation(model)
 
+function _get_all_constraint_ref_array(model::Model)
+    ref_arr = vcat([
```
Comment (blegat) [inline src/jump.jl]: Splatting with an arbitarily long vector is not a good idea, use `reduce(vcat, [...])` instead of `vcat([...]...)`

**Comment by blegat** [comment]
You will be missing `VectorOfVariables/VectorAffineFunction`-in-`Zeros/Nonpositives/Nonnegatives`.
I would recommend the following strategy:
```julia
@model(_MOIModel,
      (),                                                         # untyped scalar sets
      (MOI.EqualTo, MOI.LessThan,), #   typed scalar sets
      (),            # untyped vector sets
      (),                                                         #   typed vector sets
      (MOI.SingleVariable,),                                      # untyped scalar functions
      (MOI.ScalarAffineFunction,),                                #   typed scalar functions
      (),                                   # untyped vector functions
      ())                                #   typed vector functions
```
so that any constraint that is not in the form HyperPlane or HalfSpace is transformed into it by constraint bridges (instead of having to define a complicated iteration in LPHRepresentation as we had to do until now.
We still need to leave `SingleVariable` until https://github.com/JuliaOpt/MathOptInterface.jl/pull/659 is merged.
Then given a JuMP model, we can do
```julia
polyhedra_model = _MOIModel{Flaot64}()
bridged = MOI.Bridges.full_bridge_optimizer(polyhedra_model, Float64)
MOI.copy_to(bridged, backend(model), copy_names = false)
```
then in `polyhedra_model` you only have to deal with a fixed set of constraints that you can access with e.g. `MOI.ListOfConstraintIndices{MOI.ScalarAffineFunction{Float64}, MOI.LessThan{Float64}}()`.

File: src/jump.jl [hunk 2]
```diff
@@ -8,18 +8,143 @@ Note that if non-linear constraint are present in the model, they are ignored.
...
+
+function _get_all_constraint_array(model::Model)
+    constraint_arr = constraint_object.(
+        _get_all_constraint_ref_array(model)
+    )
+end
+
+function _get_var_bounds_array(model::Model)
+    constraint_arr = _get_all_constraint_array(model)
+    var_bounds = filter(c -> c.func isa VariableRef, constraint_arr)
```
Comment (blegat) [inline src/jump.jl]: You will be missing `VectorOfVariables`-in-`Zeros/Nonpositives/Nonnegatives`.

File: src/jump.jl [hunk 3]
```diff
@@ -8,18 +8,143 @@ Note that if non-linear constraint are present in the model, they are ignored.
...
+    return var_bounds
+end
+
+function _get_aff_constr_array(model::Model)
+    constraint_arr = _get_all_constraint_array(model)
+    linconstr = filter(c -> c.func isa GenericAffExpr, constraint_arr)
```
Comment (blegat) [inline src/jump.jl]: You will be missing `VectorAffineFunction`-in-`Zeros/Nonpositives/Nonnegatives`.

**Comment by OliverEvans96** [comment]
@blegat thanks for the detailed review!

> I would recommend the following strategy: ...

Unfortunately, I don't have any idea what you're talking about in that comment.

**Comment by blegat** [comment]
The LPHRepresentation defined here:
https://github.com/JuliaPolyhedra/Polyhedra.jl/blob/fc51bb5efdb4270ef8ff71561bf07bedc63ece0d/src/lphrep.jl#L4-L17
Is rather complicated and its design inherits from the MathProgBase LPQP model.
As a consequence, defining the `halfspaces` and `hyperplanes` iterators is rather tedious.
If we defined the model as in
https://github.com/JuliaPolyhedra/Polyhedra.jl/pull/155#issuecomment-471357070
we can simply do
```julia
struct LPHRep{T}
    model::_MOIModel{T}
end
```
The full dimension of `p::LPHRep` is simply `MOI.get(p.model, MOI.NumberOfVariables())` and the number of halfspaces is simply `MOI.get(p.model, MOI.NumberOfConstraints{MOI.ScalarAffineFunction{Float64}, MOI.LessThan{Float64}}())`, and so on.
With this representation of LPHRep, it is easy to build it form a JuMP model as detailed in
https://github.com/JuliaPolyhedra/Polyhedra.jl/pull/155#issuecomment-471357070
and it is easy to defined halfspace and hyperplanes iterators as shown above

**Comment by OliverEvans96** [comment]
Ah, thanks for the explanation - that makes sense!

File: src/jump.jl (hunk 1 above)
Comment (OliverEvans96) [reply src/jump.jl]: Very good point - thanks! Fixed.

File: src/jump.jl (hunk 2 above)
Comment (OliverEvans96) [reply src/jump.jl]: Updated.

File: src/jump.jl (hunk 3 above)
Comment (OliverEvans96) [reply src/jump.jl]: Updated.

**Comment by blegat** [comment]
I created a PR implementing the approach I detailed here: https://github.com/JuliaPolyhedra/Polyhedra.jl/pull/157.
The issue is a bit tricky and requires a good understanding of Polyhedra internals that I didn't sufficiently document, and I'd like to release it before going on holidays.
Thank for your contribution, don't hesitate to propose other changes. Sorry for taking over, I should have communicated it more clearly, e.g. by assign myself to the issue, to avoid duplicate work.

**Comment by OliverEvans96** [comment]
No worries - I trust that your solution is superior to my hack! Thanks for making the update! Let me know if there's anything else related to this PR that I can do to help.

Oliver

**Pull request closed without merging**